### ISSUE TITLE
[HunyuanImage-2.1-Distilled-Diffusers] OOM in transformer due to native SDPA decomposition 

### ISSUE THREAD

**Issue by kamalrajkannan78** [tenstorrent/tt-xla#4780]
### Describe the bug

- On 4 chips, transformer fails with `Out of Memory: Not enough space to allocate 771162112 B DRAM buffer across 12 banks, where each bank needs to store 64266240 B, but bank size is 1071821792 B (allocated: 1014289568 B, free: 57532224 B, largest free block: 47172576 B)` - [jun1_HunyuanImage_2_1_transformer_4_chips.log](https://github.com/user-attachments/files/28458892/jun1_HunyuanImage_2_1_transformer_4_chips.log)

### Root cause analysis

- 4 chips, fp32 & bf16: OOM on a 771 MB buffer = native-SDPA attn_weight + mask add producing [1,7,5224,5224] f32 joint-attention scores. Verified: 0 SDPA composites survive in the IR (composite never engages).
- Why composite skipped: _check_sdpa_constraints ([composite_ops.py:430-444](vscode-webview://1l6oi0mrmo380rdlh5jeli9at5o1d5ilhptp9kritsbpbbn0302i/python_package/tt_torch/composite_ops.py#L430-L444)) needs all SDPA inputs bf16. fp32 → Q/K/V fail; bf16 → the attn_mask is bool (i1, built at [transformer_hunyuanimage.py:818-819](vscode-webview://1l6oi0mrmo380rdlh5jeli9at5o1d5ilhptp9kritsbpbbn0302i/venv/lib/python3.12/site-packages/diffusers/models/transformers/transformer_hunyuanimage.py#L818-L819)) → fails. Confirmed 1x1x1x5224xi1 mask in IR.
- Sharding is fine (not the bug): tensor-parallel weights + heads 28→7/chip, verified in IR. OOM is accumulated residency (~11.1 GB/12 GB; 45 graphs execute first), so it's a late joint-attention block, not the first SDPA.
- 8 chips (fp32 & bf16): different/earlier failure — 'sdy.collective_permute' op requires the same type for all operands and results. Distinct from the OOM.
- Root caused is clearly explained in below comments

### TODO

- [x] **Extract the problematic SDPA's details** — shape `[1, 28→7, 5224, 5224]`, dtype (f32 scores), the `attn_mask` dtype (`bool`/`i1`), and which block it lands in.
- [x] **Build a minimal sanity repro** — a single joint-attention call (seq 5224, 28 heads, bf16 Q/K/V + **bool** mask) that reproduces both (a) the composite being skipped and (b) the f32 score-matrix OOM, without the full 17.45B model.
- [x] **Fix on the sanity repro** — make the composite engage with a non-bf16 mask: relax `_check_sdpa_constraints` to gate only on Q/K/V (or cast the mask to bf16 additive), and confirm the fused TTNN SDPA is used.
- [x] **Verify the fix** — score matrix no longer materialized; OOM cleared (and check it doesn't just shift downstream).
- [x] **Raise the PR** (and link the companion tt-mlir change if the fused-mask path needs work there).

### Logs

- [jun1_HunyuanImage_2_1_transformer_8_chips_bfp16.log](https://github.com/user-attachments/files/28458987/jun1_HunyuanImage_2_1_transformer_8_chips_bfp16.log)
- [jun1_HunyuanImage_2_1_transformer_8_chips.log](https://github.com/user-attachments/files/28458986/jun1_HunyuanImage_2_1_transformer_8_chips.log)
-  [jun1_HunyuanImage_2_1_transformer_4_chips_bfp16.log.zip](https://github.com/user-attachments/files/28458916/jun1_HunyuanImage_2_1_transformer_4_chips_bfp16.log.zip)

**Comment by kamalrajkannan78** [comment]
## 10/06/2026

### Update: root-caused the unfused SDPA (reproduced with a minimal sanity test)

Sanity repro = SDPA : Q/K/V `[1, 28, 5224, 128]` bf16, **bool** mask `[1, 1, 1, 5224]`. - [jun10_sdpa_bfp16.log](https://github.com/user-attachments/files/28803993/jun10_sdpa_bfp16.log), [test_sdpa.py](https://github.com/user-attachments/files/28803992/test_sdpa.py)

1. **Composite skipped even in bf16.** `_check_sdpa_constraints` required *every* SDPA input — including the mask — to be bf16. The model's mask is **bool**, so the composite never engaged and the native decomposition materialized the `[·, ·, 5224, 5224]` **f32 score matrix → OOM**.

2. **Relaxing the gate to Q/K/V only exposed a shape constraint.** The composite engaged but the fused op failed:
   ```
   'ttir.scaled_dot_product_attention' op Attention mask at dim 2 must match query sequence length
   ```
   The TTIR/TTNN SDPA verifier allows broadcasting the mask on dim0/dim1 but **requires dim2 == query seq len**; our mask is `[1,1,1,5224]` (dim2 = 1). So this is enforced in tt-mlir, not only the kernel - [jun10_sdpa_bfp16_after_fix_attn_issue_ttir.log](https://github.com/user-attachments/files/28804073/jun10_sdpa_bfp16_after_fix_attn_issue_ttir.log)

3. **Composite fixes (tt-xla) made single-chip pass — but the model still OOM'd.** In `composite_scaled_dot_product_attention`: **(a)** convert the bool mask → **additive bf16** with `where(mask, 0, -inf)`, and **(b)** **expand dim2** to seq_len.
   - *Why `0`/`-inf`:* the fused kernel computes `softmax(QKᵀ·scale + mask)` — it **adds** the mask before softmax, so the mask must be additive. To **keep** a key, add `0` (score unchanged → normal weight); to **drop** one, add `-inf` (`exp(-inf) = 0` → zero softmax weight). The default bool→bf16 cast (`True→1.0`, `False→0.0`) is wrong both ways: `False→0.0` adds nothing so masked keys are **not** suppressed (padding leaks in), and `True→1.0` spuriously shifts the kept scores. Measurable: PCC 0.9965 → 0.9996.
   - **(3b) CPU equivalence check** (`test_expand_dim_attn_sdpa.py`): the dim2-expand is exact — same Q/K/V, broadcast mask `[1,1,1,5224]` vs expanded `[1,1,5224,5224]` produce **bit-identical** SDPA outputs (`torch.equal=True`, `max_abs_diff=0.0`). So the expand is semantics-preserving independent of the device. - [jun10_expand_dim_spda_fix_cpu.log](https://github.com/user-attachments/files/28804822/jun10_expand_dim_spda_fix_cpu.log), [test_expand_dim_attn_sdpa.py](https://github.com/user-attachments/files/28804823/test_expand_dim_attn_sdpa.py)
   - Single-chip sanity then fused and passed; the model still OOM'd. The difference: **single-chip sanity has no sharding. - [jun10_sdpa_bfp16_after_fix_check.log](https://github.com/user-attachments/files/28804090/jun10_sdpa_bfp16_after_fix_check.log), [jun10_HunyuanImage_2_1_transformer_bfp16_4_chips_after_try_fix.log.zip](https://github.com/user-attachments/files/28804107/jun10_HunyuanImage_2_1_transformer_bfp16_4_chips_after_try_fix.log.zip) , [jun10_sdpa_bfp16_after_fix_check_without_additive_bfp16.log](https://github.com/user-attachments/files/28804380/jun10_sdpa_bfp16_after_fix_check_without_additive_bfp16.log) 

4. **Reproduced under sharding.** A multi-chip sanity (tensor-parallel, heads 28→7) does **not** fuse — SDPA is decomposed at the SHLO→TTIR stage. Root cause: under Shardy, `FlattenOrConvertComposites` flattens any composite **without a custom sharding rule** into its decomposition (RMS norm is registered; SDPA wasn't), and the resulting post-shardy `custom_call` had no SHLO→TTIR lowering. [jun10_sdpa_bfp16_mc.log](https://github.com/user-attachments/files/28804130/jun10_sdpa_bfp16_mc.log), [test_sdpa_mc.py](https://github.com/user-attachments/files/28804129/test_sdpa_mc.py)

5. **Fix:** tt-mlir PR https://github.com/tenstorrent/tt-mlir/pull/8759 — registers SDPA in the custom-sharding list + a head-sharding rule, and adds the `custom_call → ttir.scaled_dot_product_attention` lowering.
   - **Sanity (multi-chip):** sharded SDPA now fuses — `ttnn.scaled_dot_product_attention`, 0 decomposition, PCC 0.9996.
   - **Model:** 37 fused `ttnn.scaled_dot_product_attention`, **0** native decompositions — the `[·, ·, 5224, 5224]` f32 score matrices (the 771 MB OOM buffer) are eliminated, and the transformer now runs ~37 attention blocks deep instead of OOMing at the very first joint-attention block. - [jun10_HunyuanImage_2_1_transformer_bfp16_4_chips_after_f_fix1.log.zip](https://github.com/user-attachments/files/28804237/jun10_HunyuanImage_2_1_transformer_bfp16_4_chips_after_f_fix1.log.zip), [jun10_sdpa_bfp16_mc_after_f_fix1.log](https://github.com/user-attachments/files/28804229/jun10_sdpa_bfp16_mc_after_f_fix1.log)

6. **Follow-up:** with SDPA fused, the 771 MB score matrix is gone and the model runs far deeper (~37 attention blocks vs dying at block 1), but now OOMs on a small downstream buffer (`ttnn.neg` in rotary, ~117 MB, at ~97% DRAM). Will track these in separate ticket


### Repro

```
git checkout kkannan/jun10_hunyuan_sdpa_OOM
git submodule update --init --recursive

pytest  tests/torch/models/HunyuanImage_2_1/test_transformer.py -svv 
pytest tests/torch/ops/test_sdpa.py -svv
pytest tests/torch/ops/test_sdpa_mc.py -svv
```

## SDPA fusion — follow-up PRs

- [x] **tt-mlir** — support a broadcast (`dim2 == 1`) attention mask in SDPA, fixing
      `'ttir.scaled_dot_product_attention' op Attention mask at dim 2 must match query sequence length`:
  - [x] Relax the `dim2 == query seq len` check in the `ttir` + `ttnn` SDPA verifiers to also allow `dim2 == 1`.
  - [x] In the TTIR→TTNN SDPA lowering, materialize the broadcast (expand mask `dim2` → seq_len) before the kernel.
- [ ] **tt-xla** — Catch the SDPA composite for the model's mask:
  - [ ] Relax `_check_sdpa_constraints` to gate only **Q/K/V** on bf16 (not the mask).
  - [ ] Convert bool mask → **additive bf16** (`where(mask, 0, -inf)`) in `composite_scaled_dot_product_attention`.

> Order: land the tt-mlir PR first; once it accepts a `dim2 == 1` mask, drop the temporary dim2-expand from tt-xla.
